tests/unit/src/main/scala/tests/TestingServer.scala

```diff
@@ -652,6 +653,36 @@ final class TestingServer(
       _ = Assertions.assertNoDiff(folded, expected)
     } yield ()
 
+  def retrieveRanges(
+      filename: String,
+      expected: String
+  ): Future[ju.List[l.SelectionRange]] = {
+    val path = toPath(filename)
+    val input = path.toInputFromBuffers(buffers)
+    val offset = expected.indexOf("@@")
+    if (offset < 0) sys.error("missing @@")
+    val start = input.text.indexOf(expected.replace("@@", ""))
+    val point = start + offset
+    val pos = m.Position.Range(input, point, point)
+    val params =
+      new l.SelectionRangeParams(
+        path.toTextDocumentIdentifier,
+        List(pos.toLSP.getStart).asJava
+      )
+
+    server.selectionRange(params).asScala
+  }
+
+  def assertSelectionRanges(
```

Comment:
What about a recursive function:
```scala
  def assertSelectionRanges(     
      filename: String,
      ranges: List[l.SelectionRange],
      expected: *String,
  ): Unit = { expected.headOption.foreach { expectedFile =>
        val edits = RangesTextEdits.fromSelectionRanges(ranges)
        val edited = TextEdits.applyEdits(textContents(filename), edits)
       Assertions.assertNoDiff(edited, expectedFile)
      assertSelectionRanges(filename, ranges.map(_.getParent), expected.tail)
    }
  }

```
I think would look nicer in the test.

Reply:
Or even we could put all ranges in a single file the same as the FoldingRangeSuite does it

Reply:
Ahh I like the recursive function idea. That will make it much cleaner. Let me play around with this a bit.